### PR TITLE
Fix Dialog component overflow issue

### PR DESCRIPTION
**Pull request by hmstepanek**
Previously when the Dialog contents overflowed the height of the parent, they would get cut off.
image

The Dialog configuration by its parent element and the Dialog html and css has been refactored to more concisely and explicitly express their intentions. 

Previous:
image

Current:
image

Current for soon to be new dialog (See other [PR](https://github.com/hypothesis/lms/pull/859)):
image

### CODE REVIEW

File: lms/static/scripts/frontend_apps/components/Dialog.js [hunk 1]
```diff
@@ -68,48 +67,41 @@ export default function Dialog({
...
       role="dialog"
       aria-labelledby="Dialog__title"
       onKeyDown={handleKey}
       tabIndex="0"
       ref={rootEl}
     >
-      <div
-        className="Dialog__background"
-        style={{ zIndex: zIndexScale.dialogBackground }}
-      />
```
Comment (robertknight) [inline lms/static/scripts/frontend_apps/components/Dialog.js]: The reason why the dialog background is a separate element is to enable it to be transparent without the content of the dialog inheriting the transparency. MDN [does suggest](https://developer.mozilla.org/en-US/docs/Web/CSS/opacity) an alternative way of achieving this by using the `background` property, which I'd overlooked.

File: lms/static/scripts/frontend_apps/components/Dialog.js [hunk 2]
```diff
@@ -68,48 +67,41 @@ export default function Dialog({
 
   return (
     <div
+      className={classNames({
+        Dialog__background: true,
+        [contentClass]: true,
```
[the remark this answers is not in the document]
Comment (robertknight) [reply lms/static/scripts/frontend_apps/components/Dialog.js]: Just to clarify what I'd like to see here - the `contentClass` should be moved back to the dialog's content and using the approach suggested in the review while making the per-dialog content layout customizations work as before.

File: lms/static/scripts/frontend_apps/components/Dialog.js (hunk 1 above)
Comment (hmstepanek) [reply lms/static/scripts/frontend_apps/components/Dialog.js]: ^ I've now implemented it using the background property which as far as I can tell is a better way of solving this issue.

File: lms/static/scripts/frontend_apps/components/Dialog.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (hmstepanek) [reply lms/static/scripts/frontend_apps/components/Dialog.js]: Ok, I did notice that it was previously being used to control the content of the dialog but I didn't really find a good use case of that in the code but I'm happy to change it back.

File: lms/static/styles/components/_LMSFilePicker.scss [hunk 3]
```diff
@@ -1,8 +1,8 @@
 .LMSFilePicker__dialog {
-  display: flex;
-  flex-direction: column;
-  max-height: 100vh;
-  max-width: 500px;
-  width: 80vw;
-  height: 400px;
+  /* Center the dialog over the content. */
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply lms/static/styles/components/_LMSFilePicker.scss]: I'm not sure these settings are necessary at all anymore. I think the default width of 75% looks very reasonable already. If the content exceeds the parent height then it turns into a scroll so I don't think there needs to be a max height setting on here particularly to the whole viewport as the height would not exceed the parent element. Rather than setting a particular height, allowing the default height to auto adjust based on the size of the content also seems preferable to me (although this does cause the content to jump up when the error details are expanded and the content was smaller than the size of the parent due to the margin:auto where the content is centered within the parent element). I don't see that the flex settings add any value here either. For these reasons I've removed these two classes entirely. Let me know if you think otherwise!

File: lms/static/styles/components/_URLPicker.scss [hunk 4]
```diff
@@ -1,7 +0,0 @@
-.URLPicker__dialog {
-  display: flex;
-  flex-direction: column;
-  width: 80vw;
-  max-width: 500px;
```
Comment (robertknight) [inline lms/static/styles/components/_URLPicker.scss]: The `max-width: 500px` (and for `LMSFilePicker` too) were just a subjective judgement that it looked nicer to avoid having the dialog be too wide.

File: lms/static/styles/components/_Dialog.scss [hunk 5]
```diff
@@ -1,32 +1,22 @@
...
-  width: 100%;
-  height: 100%;
-}
+  overflow: scroll;
+  background-color: rgba(0, 0, 0, 0.5);
 
-.Dialog__background {
-  background-color: black;
-  bottom: 0;
+  /* Center the dialog over the parent's content. */
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply lms/static/styles/components/_Dialog.scss]: True is doesn't necessarily center the content but it does place the dialog over the content which is what I was trying to get at here. For me (not being well versed in css) it's not so obvious what it does which is why I have the comment. I'll see if I can re-word it.

File: lms/static/styles/components/_LMSFilePicker.scss [hunk 6]
```diff
@@ -1,8 +0,0 @@
-.LMSFilePicker__dialog {
-  display: flex;
-  flex-direction: column;
-  max-height: 100vh;
-  max-width: 500px;
-  width: 80vw;
-  height: 400px;
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply lms/static/styles/components/_LMSFilePicker.scss]: Ah ok that makes sense.

File: lms/static/styles/components/_URLPicker.scss (hunk 4 above)
Comment (hmstepanek) [reply lms/static/styles/components/_URLPicker.scss]: Ah ok I can add that back in with a comment explaining that choice.

File: lms/static/styles/components/_Dialog.scss [hunk 7]
```diff
@@ -1,32 +1,22 @@
-.Dialog__container {
+.Dialog__background {
   display: flex;
-  flex-direction: column;
   justify-content: center;
```
[the remark this answers is not in the document]
Comment (hmstepanek) [reply lms/static/styles/components/_Dialog.scss]: According to [this guide](https://css-tricks.com/centering-css-complete-guide/) you need both justify-content and align-items set to center to center vertically and horizontally. 
image
I'm also finding this to be true just experimenting with it locally.

[the remark this answers is not in the document]
Comment (hmstepanek) [reply lms/static/styles/components/_Dialog.scss]: > I mentioned in my earlier comments that vertically centering the content may be a problem if the iframe ends up being so tall for some reason that the top of the dialog ends up off-screen. If that does come to pass, one way to resolve that would be to position the background's content at the top of the dialog but add a vertical spacer (using flex-basis and flex-shrink) that stretches up to say 200px (if the content is short) or shrinks if the content is tall.

So what you're getting at here is that it might be better to have the content within the Dialog be of a minimum a certain size and of a maximum a certain size such that you can see the whole of the content without needing to scroll (aka: title to button). I believe these would need to be set by the content containing div such as ErrorDialog and not the Dialog div though. If you set them on the Dialog content div the space is added at the bottom which ends up looking something like this:
image

File: lms/static/styles/components/_URLPicker.scss (hunk 4 above)
Comment (hmstepanek) [reply lms/static/styles/components/_URLPicker.scss]: Note in order to allow overriding of the width like this I had to add margin-left and right auto. I couldn't get it to center by using flex settings. I admit don't really understand why that was the case though.

File: lms/static/styles/components/_Dialog.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (hmstepanek) [reply lms/static/styles/components/_Dialog.scss]: Note I have also added a little wrapper with the min-height set to 0 to prevent the top from overflowing upwards if the height of the content exceeds the ancestor's height as per https://css-tricks.com/flexbox-truncated-text/.

**Comment by hmstepanek** [comment]
> Just a general question. Do we care about css media queries for the LMS app? Do we need to worry about cell phones or narrow screens?

That's a good question. While I've been trying to make my changes mobile friendly but I'm not 100% sure and some of the views (like the file picker) don't seem very mobile friendly. @robertknight do you know?

**Comment by robertknight** [comment]
> Just a general question. Do we care about css media queries for the LMS app? Do we need to worry about cell phones or narrow screens?

We should try to make the UI responsive in general, but from what I've seen, we only need to worry about desktop and tablet form factors for the LMS UI. If you go to https://hypothesis.instructure.com on a typical phone for example, you can see that the Canvas UI itself is responsive in some respects but is not designed to be used on a screen that small. IMO this means we can usually be responsive enough with just flexbox/grid layouts rather than needing to switch to a completely different layout using media queries.

**Comment by robertknight** [comment]
The **Enter URL** dialog is now larger than it should be:

Screenshot 2019-08-27 16 46 00

File: lms/static/styles/components/_Dialog.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (robertknight) [reply lms/static/styles/components/_Dialog.scss]: > According to this guide you need both justify-content and align-items set to center vertically and horizontally.

Right, sorry. That's correct. Flexbox lays items out in either a vertical or horizontal line. `justify-content` then controls the alignment of an item in the direction that the line is laid out and `align-items` controls the alignment in the perpendicular direction.

File: lms/static/styles/components/_URLPicker.scss [hunk 8]
```diff
@@ -1,7 +1,6 @@
 .URLPicker__dialog {
-  display: flex;
-  flex-direction: column;
-  width: 80vw;
+  /* Prevent the Dialog from jumping size when changing state from fetching to fetched. */
```
Comment (robertknight) [inline lms/static/styles/components/_URLPicker.scss]: `URLPicker` is the "Enter URL" dialog (see `URLPicker.js`), it doesn't have any fetching/fetched state associated with it.

File: lms/static/styles/components/_Dialog.scss [hunk 9]
```diff
@@ -70,3 +72,9 @@
     }
   }
 }
+
+.Dialog__dimension {
+  /* See https://css-tricks.com/flexbox-truncated-text/ */
+  min-height: 0;
+  width: 75%;
```
Comment (robertknight) [inline lms/static/styles/components/_Dialog.scss]: I understand the linked article, but I'm not clear on what you are trying to do here. Can you clarify?

File: lms/static/styles/components/_LMSFilePicker.scss [hunk 10]
```diff
@@ -1,8 +1,6 @@
 .LMSFilePicker__dialog {
-  display: flex;
-  flex-direction: column;
-  max-height: 100vh;
+  /* Prevent the Dialog from jumping size when changing state from fetching to fetched. */
+  min-height: 400px;
```
Comment (robertknight) [inline lms/static/styles/components/_LMSFilePicker.scss]: I found a regression here. In production, the dialog will shrink if the iframe is vertically smaller than ~500px:

Screenshot 2019-08-27 17 08 35

Locally that doesn't work:

Screenshot 2019-08-27 17 01 46

This is what `max-height: 100vh` achieved previously. That caused problems when the content height exceeded `100vh` though.

File: lms/static/styles/components/_Dialog.scss [hunk 11]
```diff
@@ -1,33 +1,30 @@
...
+  display: flex;
+  flex-direction: column;
   background-color: white;
   border-radius: 3px;
-  margin-left: auto;
-  margin-right: auto;
   padding: 20px;
+  /* Center the content horizontally in case it's width is overriden. */
+  margin-right: auto;
+  margin-left: auto;
```
Comment (robertknight) [inline lms/static/styles/components/_Dialog.scss]: Wait, why is this needed? If the content is centered it shouldn't matter how wide it is.

Presumably it relates to this comment - https://github.com/hypothesis/lms/pull/927/files#r317419169?

**Comment by hmstepanek** [comment]
I'm going to close this for now. As discussed, @robertknight would like to look at this in more detail when he has some free time.
